### ISSUE TITLE
py_check и валидация с нормализацией

### ISSUE THREAD

**Issue by socketpair** [ponyorm/pony#152]
На данный момент, py_check возвращает `True`/`False`, Это уже отлично, но не покрывает вариант с нормализацией данных. Вот примеры использования:

1. Передали строку с датой, а там 32-й день месяца. Такое может оказаться, когда к дату после прибавления дня не нормализировали. Некоторые программы (в т.ч. SQL-движки) принимают такую timetuple и нормализируют молча сами, кто-то так и записывает в БД, генеря ошибки потом, а кто-то сразу выдаёт ошибку.
2. допустим, маску подсети в нашей бд мы храним как число ненулевых бит, тогда можно разрешить вставлять записи  в любом виде (CIDR Или в числе бит) но в БД должно писаться в виде числа бит.
3. дата в БД хранится как дата, но в качестве данных передали timestamp. мы сможем в этой функции написать конвертор, что удобно.

и т.д.

Другими словами, при указании значения надо бы чтобы вызывалась другая функция наподобие `py_check`, но она должна вернуть либо новое, более правильное значение, которое будет использоваться вместо оригинального, либо должна бросить ЛЮБОЕ исключение (типа ValueError), либо вернут оригинальное значение (свой аргумент).

Видимо, придётся разрешить использовать только либо `py_check` либо `py_normalize_and_validate` (или как вы посчитаете нужным её назвать). иначе не ясен смысл одновременного использования.

В FAQ нужно написать про сюрпризы когда логика программы такова, что, вставив значение в БД, она полагает, что потом при чтении именно это значение она и получит. Здесь важно ещё что тип данных может смениться.

**Comment by Vitalium** [comment]
Для этого возможно использовать методы before_update, before_create, etc, где можно нормализовать все поля, если это необходимо.

**Comment by andreymal** [comment]
Нельзя, ибо в before_create/update скормить ненормализованное значение неоткуда (в самом объекте ненормализованное хранить невозможно — ORM матерится ValueError'ом)

**Comment by Vitalium** [comment]
Да, так и есть :(
Конечно было бы неплохо добавить такие вещи как validate_field и filter в WTForms.
Хотелось знать, есть ли это в планах.

**Comment by andreymal** [comment]
Лично я данные перед понями прогоняю через Cerberus и мне норм)

**Comment by ml31415** [comment]
I totally agree, it would be more than useful, if `py_check` could also modify the value itself. Furthermore, right now there is an even more serious problem. `py_check` is not even executed, when no converters are set for a certain data type. `val` is simply returned without a check in https://github.com/ponyorm/pony/blob/orm/pony/orm/core.py#L2026
